### PR TITLE
feat(useStyles): add caching when no inline overrides are applied

### PR DESCRIPTION
**Pull request by mnajdova**
This PR is adding styles caching when the `useStyles` hook is used and **no inline overrides** are applied on the component (styles, design and variables props are not used). The caching is containing the following:
- theme (weak map referenced key)
- display name
- slot name
- props used in the styles
- rtl mode
- disable animations mode

We can enable the caching only when the `useStyles` is used at this moment, because by using it, we are sure that all props sent to the styles function are primitive values, which can be easily cached.

**BREAKING CHANGE**

The caching for both styles and variables is enabled by default. If you need to opt out of this, you can use the new `performance` prop on the `Provider`:

```
<Provider theme={theme} performance={{ enableStylesCaching: false }}>
  ...
</Provider>
```

Note that the variables cache was enabled before this PR, so it is not considered as a breaking change. However you can disabled that one if needed:

```
<Provider 
  theme={theme} 
  performance={{ 
    enableStylesCaching: false,
    enableVariablesCaching: false 
  }}
>
  ...
</Provider>
```

### CODE REVIEW

File: packages/styles/src/types.ts [hunk 1]
```diff
@@ -188,6 +188,8 @@ export interface ComponentSlotStylesInput<TProps = {}, TVars = {}>
 export interface ComponentSlotStylesPrepared<TProps = {}, TVars = {}>
   extends Record<string, ComponentSlotStyleFunction<TProps, TVars>> {}
 
+export interface ComponentSlotStylesResolved extends Record<string, ICSSInJSStyle> {}
```
Comment (layershifter) [inline packages/styles/src/types.ts]: Can we please extract this change to a separate PR? It's absolutely valid, but it creates noise in this PR

File: packages/react-bindings/src/styles/resolveStylesAndClasses.tsx [hunk 2]
```diff
@@ -3,71 +3,158 @@ import {
...
           delete resolvedStyles[slotName]['_debug']
         }
 
-        return resolvedStyles[cacheKey]
+        return resolvedStyles[lazyEvaluationKey]
       },
     })
 
-    Object.defineProperty(classes, slotName, {
+    // TODO: Fix this if possible
```
Comment (mnajdova) [inline packages/react-bindings/src/styles/resolveStylesAndClasses.tsx]: No idea how to fix this at this moment :(

File: packages/react-bindings/test/styles/resolveStylesAndClasses-test.ts [hunk 3]
```diff
@@ -52,17 +54,83 @@ describe('resolveStylesAndClasses', () => {
...
+
+    expect(resolvedStyles.root).toMatchObject({ color: 'red' })
+    expect(componentStyles.root).toHaveBeenCalledTimes(1)
+    expect(secondResolvedStyles.root).toMatchObject({ color: 'red' })
+    expect(componentStyles.root).toHaveBeenCalledTimes(1)
+  })
+
+  test('caches classes for no props', () => {
+    const renderStyles = jest.fn().mockReturnValue('a')
+    const { classes } = resolveStylesAndClasses(componentStyles, styleParam, renderStyles, true, 'Test1', theme, {})
```
Comment (mnajdova) [inline packages/react-bindings/test/styles/resolveStylesAndClasses-test.ts]: I have to change the display name here in order to avoid colliding of the cache with previous examples. Other options is to provide different theme (maybe this is better), or to add some option for invalidating the cache...

File: packages/react-bindings/src/hooks/useStyles.ts [hunk 4]
```diff
@@ -81,7 +80,7 @@ const useStyles = <StyleProps extends PrimitiveProps>(
     rtl,
     saveDebug: fluentUIDebug => (debug.current = { fluentUIDebug }),
     theme: context.theme,
-    _internal_resolvedComponentVariables: context._internal_resolvedComponentVariables,
+    __experimental_cache: true,
```
Comment (mnajdova) [inline packages/react-bindings/src/hooks/useStyles.ts]: We can add configuration for whether the caching should be enabled if needed, for now it is always true when using the useStyles hook

File: packages/react/src/components/Provider/Provider.tsx [hunk 5]
```diff
@@ -181,6 +184,7 @@ class Provider extends React.Component<WithAsProp<ProviderProps>> {
       theme,
       rtl,
       disableAnimations,
+      enableCaching,
```
Comment (mnajdova) [inline packages/react/src/components/Provider/Provider.tsx]: Maybe the value should be `disableCaching` as we already have `disableAnimations` in the context.. But it feels wrong, as I want to disable it by default...

File: packages/react/src/components/Provider/Provider.tsx [hunk 6]
```diff
@@ -33,6 +33,7 @@ export interface ProviderProps extends ChildrenComponentProps {
   renderer?: Renderer
   rtl?: boolean
   disableAnimations?: boolean
+  enableCaching?: boolean
```
Comment (layershifter) [inline packages/react/src/components/Provider/Provider.tsx]: 1. Let's **enable** it by default as it's our strategy to have caching
2. Let's use bitwise operations for flags (https://blog.rinatussenov.com/juggling-bits-in-javascript-bitmasks-128ad5f31bed) as we obviously want to add more settings like caching variables and disabling Fela plugins:

```tsx
<Provider performance={PERFORMANCE_USE_CACHE | PERFORMANCE_USE_VARIABLES_CACHE | PERFORMANCE_NO_EXPAND_CSS} />
```

File: packages/react-bindings/src/styles/getStyles.ts [hunk 7]
```diff
@@ -42,81 +43,90 @@ export type GetStylesResult = {
   theme: StylesContextValue['theme']
 }
 
+const variablesCache = new WeakMap<ThemePrepared, Record<string, any>>()
+
+const resolveVariables = (
+  displayName: string,
+  theme: ThemePrepared,
+  variables: any | undefined,
+): any => {
```
Comment (layershifter) [inline packages/react-bindings/src/styles/getStyles.ts]: Oops, let's fix any there:

```suggestion
): ComponentVariablesObject => {
```

File: packages/react-bindings/src/styles/getStyles.ts [hunk 8]
```diff
@@ -42,81 +43,90 @@ export type GetStylesResult = {
   theme: StylesContextValue['theme']
 }
 
+const variablesCache = new WeakMap<ThemePrepared, Record<string, any>>()
+
+const resolveVariables = (
+  displayName: string,
+  theme: ThemePrepared,
+  variables: any | undefined,
```
Comment (layershifter) [inline packages/react-bindings/src/styles/getStyles.ts]:
```suggestion
  variables: ComponentVariablesInput | undefined,
```

File: packages/react-bindings/src/styles/getStyles.ts [hunk 9]
```diff
@@ -42,81 +43,90 @@ export type GetStylesResult = {
...
+  //
+  // Simple caching model, works only if there is no `props.variables`
+  // Resolves variables for this component, cache the result in provider
+  //
+
+  if (!variablesCache.has(theme)) {
+    variablesCache.set(theme, {})
+  }
+
+  // @ts-ignore
```
Comment (layershifter) [inline packages/react-bindings/src/styles/getStyles.ts]:
```suggestion
```

File: packages/react-bindings/src/styles/getStyles.ts [hunk 10]
```diff
@@ -42,81 +43,90 @@ export type GetStylesResult = {
...
+  if (!variablesCache.get(theme)[displayName]) {
+    // component variables must not be undefined/null (see mergeComponentVariables contract)
+    variablesCache.set(theme, {
+      ...variablesCache.get(theme),
+      [displayName]: callable(theme.componentVariables[displayName])(theme.siteVariables) || {},
+    })
```
Comment (layershifter) [inline packages/react-bindings/src/styles/getStyles.ts]: Are we sure that it's the best option?

File: packages/react-bindings/src/styles/getStyles.ts [hunk 11]
```diff
@@ -42,81 +43,90 @@ export type GetStylesResult = {
...
+  }
+
+  if (variables === undefined) {
+    // @ts-ignore
```
Comment (layershifter) [inline packages/react-bindings/src/styles/getStyles.ts]:
```suggestion
```

File: packages/react-bindings/src/styles/getStyles.ts [hunk 12]
```diff
@@ -42,81 +43,90 @@ export type GetStylesResult = {
...
+    return variablesCache.get(theme)[displayName]
+  }
+
+  return mergeComponentVariables(
+    // @ts-ignore
```
Comment (layershifter) [inline packages/react-bindings/src/styles/getStyles.ts]:
```suggestion
```

File: packages/react-bindings/src/styles/getStyles.ts [hunk 13]
```diff
@@ -42,81 +43,90 @@ export type GetStylesResult = {
...
+  // - resolve variables (siteVariables => componentVariables + props.variables)
+  // - resolve styles (with resolvedVariables & props.styles & props.design)
+  // - compute classes (with resolvedStyles)
+  //
 
-  // Resolve styles using resolved variables, merge results, allow props.styles to override
-  let mergedStyles: ComponentSlotStylesPrepared = theme.componentStyles[displayName] || {
-    root: () => ({}),
-  }
+  const resolvedVariables: any = resolveVariables(displayName, theme, props.variables)
```
Comment (layershifter) [inline packages/react-bindings/src/styles/getStyles.ts]:
```suggestion
  const resolvedVariables = resolveVariables(displayName, theme, props.variables)
```

File: packages/react-bindings/src/styles/getStyles.ts [hunk 14]
```diff
@@ -42,81 +43,90 @@ export type GetStylesResult = {
...
+    cacheEnabled: enableCaching && noInlineOverrides,
+    styleProps: restProps,
+  })
 
   // conditionally add sources for evaluating debug information to component
   if (process.env.NODE_ENV !== 'production' && isDebugEnabled) {
     saveDebug({
       componentName: displayName,
       componentVariables: _.filter(
+        // @ts-ignore
```
Comment (layershifter) [inline packages/react-bindings/src/styles/getStyles.ts]:
```suggestion
```

File: packages/react-bindings/src/styles/resolveStylesAndClasses.tsx [hunk 15]
```diff
@@ -3,71 +3,158 @@ import {
...
 import { ComponentSlotClasses } from '../styles/types'
 
-// Both resolvedStyles and classes are objects of getters with lazy evaluation
+export type ResolveStylesResult = {
+  resolvedStyles: ComponentSlotStylesResolved
+  resolvedStylesDebug: Record<string, { styles: Object }[]>
+  classes: ComponentSlotClasses
+}
+
+// TODO: Describe me
```
Comment (layershifter) [inline packages/react-bindings/src/styles/resolveStylesAndClasses.tsx]: ???

File: packages/react-bindings/src/styles/resolveStylesAndClasses.tsx [hunk 16]
```diff
@@ -3,71 +3,158 @@ import {
...
+    }
+  }
+
   Object.keys(mergedStyles).forEach(slotName => {
     // resolve/render slot styles once and cache
-    const cacheKey = `${slotName}__return`
+    const lazyEvaluationKey = `${slotName}__return`
+    const stylesCacheKey =
+      cacheEnabled && displayName && props
+        ? `${displayName}: ${slotName}: ${JSON.stringify(props)}${styleParam.rtl}${
```
Comment (layershifter) [inline packages/react-bindings/src/styles/resolveStylesAndClasses.tsx]: Only `slotName` is changed there. We can move out this from `forEach` as it really no effective to run it every time:

```tsx
const componentCacheKey = `${displayName}:${JSON.stringify(props)}${styleParam.rtl}${styleParam.disableAnimations}`

// 
// forEach
const slotCacheKey = slotName + stylesCacheKey
```

File: packages/react-bindings/src/styles/resolveStylesAndClasses.tsx [hunk 17]
```diff
@@ -3,71 +3,158 @@ import {
...
+
+        resolvedStyles[lazyEvaluationKey] = val
       },
-      get() {
-        if (resolvedStyles[cacheKey]) {
-          return resolvedStyles[cacheKey]
+      get(): ICSSInJSStyle {
+        // If caching enabled and entry exists, get from cache, avoid lazy evaluation
+        if (cacheEnabled && theme) {
+          // @ts-ignore
```
Comment (layershifter) [inline packages/react-bindings/src/styles/resolveStylesAndClasses.tsx]:
```suggestion
```

File: packages/react-bindings/src/styles/resolveStylesAndClasses.tsx [hunk 18]
```diff
@@ -3,71 +3,158 @@ import {
...
+          if (stylesCache.get(theme)[stylesCacheKey]) {
+            // @ts-ignore
```
Comment (layershifter) [inline packages/react-bindings/src/styles/resolveStylesAndClasses.tsx]:
```suggestion
```

File: packages/react-bindings/src/styles/resolveStylesAndClasses.tsx [hunk 19]
```diff
@@ -3,71 +3,158 @@ import {
...
   renderStyles: (styles: ICSSInJSStyle) => string,
-): {
-  resolvedStyles: ICSSInJSStyle
-  resolvedStylesDebug: Record<string, { styles: Object }[]>
-  classes: ComponentSlotClasses
-} => {
+  cacheEnabled?: boolean | undefined,
+  displayName?: string,
+  theme?: ThemePrepared,
+  props?: object,
```
Comment (layershifter) [inline packages/react-bindings/src/styles/resolveStylesAndClasses.tsx]: Can we somehow reduce the amount of passed things?

File: packages/react-bindings/src/styles/resolveStylesAndClasses.tsx (hunk 16 above)
Comment (miroslavstastny) [inline packages/react-bindings/src/styles/resolveStylesAndClasses.tsx]: `JSON.stringify` will not work as expected as the output depends on key insertion order.

```js
JSON.stringify({a: 1, b:2}) !== JSON.stringify({b: 2, a:1})
```

Comment (mnajdova) [reply packages/react-bindings/src/styles/resolveStylesAndClasses.tsx]: We are always sending the props in the same order in useStyles, that's why we can use JSON.stringify. See some example, for the checkbox component:
```
  const { classes, styles: resolvedStyles } = useStyles(Checkbox.displayName, {
    className: Checkbox.className,
    mapPropsToStyles: () => ({
      checked: state.checked,
      disabled,
      labelPosition,
      toggle,
    }),
    mapPropsToInlineStyles: () => ({
      className,
      design,
      styles,
      variables,
    }),
    rtl: context.rtl,
  })
```
the order will always be 
```
      checked,
      disabled,
      labelPosition,
      toggle,
```

Comment (miroslavstastny) [reply packages/react-bindings/src/styles/resolveStylesAndClasses.tsx]: 🤦‍♂ my bad

File: packages/react-bindings/src/styles/getStyles.ts (hunk 10 above)
Comment (mnajdova) [reply packages/react-bindings/src/styles/getStyles.ts]: Refactored

File: packages/react-bindings/src/styles/resolveStylesAndClasses.tsx (hunk 19 above)
Comment (mnajdova) [reply packages/react-bindings/src/styles/resolveStylesAndClasses.tsx]: We can extract these to config object, but it seems awkward to have part of the properties first line arguments, and than part of them in config object.. The other options is to move everything to config object - maybe this one is better, what do you think?

File: packages/react/src/components/Provider/Provider.tsx (hunk 6 above)
Comment (mnajdova) [reply packages/react/src/components/Provider/Provider.tsx]: 1. Will enable it by default.
2. We can't do it like that, because we are merging with above theme, so if something is not defined is not the same as if something is false. For undefined we are taking the value from the parent, for false we are overriding. Maybe we should make option performance where we can add all these different options in the future, like
`<Provider performance={{useStylesCache: true, useVariablesCache: true}}></Provider>`
Should we change it like this, or leave it as a first class prop on the Provider?

Comment (layershifter) [reply packages/react/src/components/Provider/Provider.tsx]:
```
performance={{useStylesCache: true, useVariablesCache: true}}>
```

Is okay, let's do it 👍

File: packages/react-bindings/src/hooks/useStyles.ts [hunk 20]
```diff
@@ -17,7 +17,7 @@ import {
 } from '../styles/types'
 import getStyles from '../styles/getStyles'
 
-type PrimitiveProps = Record<string, boolean | number | string | undefined>
+export type PrimitiveProps = Record<string, boolean | number | string | undefined>
```
Comment (layershifter) [inline packages/react-bindings/src/hooks/useStyles.ts]: Let's move this to `packages/react-bindings/src/styles/types.ts`

File: packages/react-bindings/src/styles/resolveStylesAndClasses.tsx (hunk 19 above)
Comment (layershifter) [reply packages/react-bindings/src/styles/resolveStylesAndClasses.tsx]: Can we merge `resolveStyles` and `resolveStylesAndClasses` functions as I don't understand the separation there anymore. I think that I will solve params issue.

File: packages/react/src/components/Provider/Provider.tsx (hunk 6 above)
Comment (mnajdova) [reply packages/react/src/components/Provider/Provider.tsx]: I've renamed tham to `enableVariablesCache` and `enableStylesCache`, as we already have disable/emable naming in the context

File: packages/react-bindings/test/styles/resolveStyles-test.ts [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+    const renderStyles = jest.fn().mockReturnValue('a')
+    const propsInlineOverrides: GetStylesOptions['props'][] = [
+      { styles: { fontSize: '10px' } },
+      { design: { left: '10px' } },
+      { variables: { backgroundColor: 'yellow' } },
+    ]
+
+    const propsInlineOverridesSize = propsInlineOverrides.length
+
+    for (let i = 0; i < propsInlineOverridesSize; i++) {
```
Comment (layershifter) [inline packages/react-bindings/test/styles/resolveStyles-test.ts]: `_.forEach()`?

File: packages/react-bindings/src/styles/resolveStyles.ts [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+        if (process.env.NODE_ENV !== 'production' && isDebugEnabled) {
+          resolvedStylesDebug[slotName] = resolvedStyles[slotName]['_debug']
+          delete resolvedStyles[slotName]['_debug']
+        }
+
+        return resolvedStyles[lazyEvaluationKey]
+      },
+    })
+
+    // const className = slotName === 'root' ? '__root' : slotName
```
Comment (layershifter) [inline packages/react-bindings/src/styles/resolveStyles.ts]: Oops

**Review by layershifter — APPROVED** [review]
